File: UI/src/main/java/ca/warp7/rt/feature/views/ViewsFeature.java
package ca.warp7.rt.feature.views;

import ca.warp7.rt.core.ft.Feature;
import ca.warp7.rt.core.ft.FeatureItemTab;
import ca.warp7.rt.core.ft.FeatureTabFactory;
import ca.warp7.rt.core.ft.FeatureUtils;
import javafx.scene.Parent;

import java.util.Arrays;
import java.util.List;

import static ca.warp7.rt.core.ft.FeatureItemTab.Group;

public class ViewsFeature implements Feature {

    private FeatureTabFactory factory = new FeatureTabFactory("fas-eye:16:gray",
            getFeatureId(), Group.WithFeature);

    private Parent preLoaded;

    @Override
    public List<FeatureItemTab> getLoadedTabs() {
        return Arrays.asList(
                factory.get("Raw Data", ""),
                factory.get("Auto List", ""),
                factory.get("Endgame", ""),
                factory.get("Team Averages", ""),
                factory.get("Cycle Matrix", "")
        );
    }

    @Override
    public String getFeatureId() {
        return "views";
    }

    @Override
    public Parent onOpenTab(FeatureItemTab tab) {
        if (preLoaded == null) {
            preLoaded = FeatureUtils.loadParent("/ca/warp7/rt/feature/views/Views.fxml");
        }
        return preLoaded;
    }
}

File: UI/src/main/kotlin/ca/warp7/rt/ext/views/ViewsFeature.kt
package ca.warp7.rt.ext.views

import ca.warp7.rt.core.feature.Feature
import ca.warp7.rt.core.feature.FeatureLink
import ca.warp7.rt.core.feature.loadParent
import javafx.scene.Parent
import javafx.scene.control.TreeItem
import javafx.scene.control.TreeView


class ViewsFeature : Feature {

    private var preLoaded: Parent? = null

    override val link = FeatureLink("Restructured Tables", "fas-database", 18)

    override fun onOpen(): Pair<Parent?, Parent?> {
        if (preLoaded == null) {
            preLoaded = loadParent("/ca/warp7/rt/ext/views/Table.fxml")
        }

        val rootItem = TreeItem("Tables")
        rootItem.isExpanded = true
        for (i in 1..5) {
            val item = TreeItem("Table $i")
            rootItem.children.add(item)
        }
        val tree = TreeView(rootItem)
        return tree to preLoaded
    }
}